artifactory/services/download.go

```diff
@@ -109,25 +102,37 @@ func (ds *DownloadService) DownloadFiles(downloadParams ...DownloadParams) (*uti
 	ds.prepareTasks(producerConsumer, expectedChan, successCounters, errorsQueue, downloadParams...)
 
 	e = ds.performTasks(producerConsumer, errorsQueue)
-	if e != nil {
-		return nil, e
-	}
 	totalSuccess := 0
 	for _, v := range successCounters {
 		totalSuccess += v
 	}
-	return ds.getOperationSummary(totalSuccess, <-expectedChan-totalSuccess), nil
+	return ds.getOperationSummary(totalSuccess, <-expectedChan-totalSuccess), e
 }
 
 func (ds *DownloadService) prepareTasks(producer parallel.Runner, expectedChan chan int, successCounters []int, errorsQueue *clientutils.ErrorsQueue, downloadParamsSlice ...DownloadParams) {
 	go func() {
 		defer producer.Done()
 		defer close(expectedChan)
 		totalTasks := 0
+		defer func() {
+			expectedChan <- totalTasks
+		}()
+		artifactoryVersionStr, err := ds.GetArtifactoryDetails().GetVersion()
+		if err != nil {
+			log.Error(err)
+			errorsQueue.AddError(err)
+			return
+		}
+		artifactoryVersion := version.NewVersion(artifactoryVersionStr)
 		// Iterate over file-spec groups and produce download tasks.
 		// When encountering an error, log and move to next group.
 		for _, downloadParams := range downloadParamsSlice {
-			var err error
+			err = utils.CheckIfVersionCompatible(downloadParams.ArtifactoryCommonParams, artifactoryVersion)
```

Comment:
This check is in the loop because the transitive option is part of the spec files, so we must iterate the files groups to check if it's true.
The reason I fetch the version outside the function is to avoid creating a new Version struct each time this function is called. On the other hand, fetching it inside the function will probably improve the code readability. Do you think it's better than the existing solution?